Development tfchain reservation (#216)
threefoldtech/jumpscaleX: Jumpscale/clients/blockchain/tfchain/README.md

        > to: 0125c0156f6c1c0bc43c7d38e17f8948300564bef63caac05c08b0fd68996e494704bbbe0268cb
        > from: 01773a1dd123347e1030f0822cb8d22082fe3f9b0ea8563d4ac8e7abc377eba920c47efb2fd736
```

### Capacity reservation

During the fist beta phase of the public launch of the TF grid, beta tester will be able to reserve 2 kind of workload on the grid.

- Zero-OS virtual machines
- S3 archive storage instances

**At the time of writing, everything happens on the testnet network**
**Don't send real TFT from the main network !!**

#### How to reserve some capacity on the Threefold Grid

To be able to make a reservation you first need to:

- have a wallet with some fund
- record a threebot on the tfchain

Once you have both, you can then use your wallet client to do a reservation.
The wallet expose a module called `capacity`. On this module you will find function to reserver the different type of capacity.

Examples:

```python
c = j.clients.tfchain.myclient
w = c.wallets.mywallet

> tx_id = w.capacity.reserve_s3(

Reviewer: I think we usually call this `txid` without the underscore. Also I would suggest to also supported the refund and source parameters, as we do in the other functionality (e.g. `send`), this allows you to choose from which wallet(s) to fund and/or which wallet to refund into if needed. This would even allow to reserve funding from a multi-sig wallet, which is otherwise not possible.

If so, you'll also need to adapt the return type, to be similar to the regular send functionality, as it might for example be possible that you cannot submit yet due to missing signature(s).

Author: >this allows you to choose from which wallet(s) to fund and/or which wallet to refund into if needed

Capacity module is on the wallet object, not the tfchain client. so you already have chosen which wallet to use a this point.

Regarding refund this is handle later on, if the grid broker fails to deploy the workload, then the grid broker does the refund.
Multi-sig is not something we support now for reservation, and I don't think we will ever need it to be honest, I don't see any use case for it at the moment

Reviewer: You get me wrong, it's about from which address to fund from your wallet, not from another wallet. So I'm talking purely about funding here. By default the funding uses all available addresses of your personal wallet. However, could be that for whatever reason you want to fund only from a specific wallet address (e.g. perhaps you want that wallet address to be used for capacity purposes, and not some address you might use for other stuff). The same feature that allows that, also automatically allows you to fund from a multi-sign wallet you own.

This is not about the public key or anything else capacity-specific, purely about funding.

Same when I talk about refunding. I am just talking about the refund coin output that you might create because the sum of your coin inputs overreaches the amount you wishes to send. Not talking about refund as in refund if capacity fails. 

This entire comment was about the funding you do for this Txn.

Author: Ok got it. Thanks for the hint. I'm not used to this blockchain logic yet :-D

    size=1, # each workload have a different size available
    email='user@email.com', # the email on which you will received the connection information
    bot_id='my3bot.example.org') # your threebot id, it can be any of the names you gave to your 3bot
```

> The result of the `reserve_s3` method call is the transaction ID in which you reservation has been created.

Reviewer: no need to use the web explorer though. You can also do it using our JSX client:

```python
j.clients.tfchain.myclient.transaction_get(txid)
```

Author: good I'll add that too

You can check it on our [explorer](https://explorer.testnet.threefoldtoken.com/) by entering the transaction ID in the `Search by hash` field of the explorer form.

> Your workload is currently being deployed, you will receive an email with the connection information within a few minutes.

Reviewer: I would say `as soon as it is ready, usually within a few minutes`.

        > to: 0125c0156f6c1c0bc43c7d38e17f8948300564bef63caac05c08b0fd68996e494704bbbe0268cb
        > from: 01773a1dd123347e1030f0822cb8d22082fe3f9b0ea8563d4ac8e7abc377eba920c47efb2fd736
```

### Capacity reservation

During the fist beta phase of the public launch of the TF grid, beta tester will be able to reserve 2 kind of workload on the grid.

- Zero-OS virtual machines
- S3 archive storage instances

**At the time of writing, everything happens on the testnet network**
**Don't send real TFT from the main network !!**

#### How to reserve some capacity on the Threefold Grid

To be able to make a reservation you first need to:

> - have a wallet with some fund

Reviewer: ```suggestion
- have a wallet with sufficient funds
```
Some is wrong, as `0.000001` TFT is also some funds, but certainly won't do you any good

- record a threebot on the tfchain

Once you have both, you can then use your wallet client to do a reservation.
The wallet expose a module called `capacity`. On this module you will find function to reserver the different type of capacity.

Examples:

```python
c = j.clients.tfchain.myclient
w = c.wallets.mywallet

> txid = w.capacity.reserve_s3(

Reviewer: ```suggestion
result = w.capacity.reserve_s3(
```

    size=1, # each workload have a different size available
    email='user@email.com', # the email on which you will received the connection information
    bot_id='my3bot.example.org') # your threebot id, it can be any of the names you gave to your 3bot
```

> The result of the `reserve_s3` method call is the transaction ID in which you reservation has been created.

Reviewer: ```suggestion
The result of the `reserve_s3` method call is a tuple containing the transaction and the submission status as a boolean.
```

You can check it on our [explorer](https://explorer.testnet.threefoldtoken.com/) by entering the transaction ID in the `Search by hash` field of the explorer form or using the tfchain client:

```python

> transaction = c.transaction_get(txid)

Reviewer: ```suggestion
transaction = c.transaction_get(result.transaction.id)
```

Only required of course if `result.submitted` is `True`
